default-fields/0000-default-fields.md

```diff
@@ -6,31 +6,268 @@
+
+```rust
+struct Foo {
+    a: &'static str,
+    b: bool = true,
+    c: i32,
+}
+
+let foo = Foo {
+    a: "Hello",
+    c: 42,
+};
+```
 
 # Detailed design
 [design]: #detailed-design
 
-This is the bulk of the RFC. Explain the design in enough detail for somebody familiar
-with the language to understand, and for somebody familiar with the compiler to implement.
-This should get into specifics and corner-cases, and include examples of how the feature is used.
+## Grammar
+
+In the initialiser for a structure, a default value expression can be optionally supplied for a field:
+
+```
+struct_field : ident ':' type_path |
+               ident ':' type_path '=' expr
+```
+
+The syntax is modeled after `const` expressions. Field defaults are only valid for classic C structures:
+
+```
+structure : 'struct' ident '{' struct_field
+                               [ ',' struct_field ] *
+                           '}' ;
+```
+
+## Interpretation
+
+The value of a field default must be a compile-time expression. So any expression that's valid as `const` can be used as a field default. This ensures values that aren't specified by the caller are deterministic and cheap to produce.
+
+Valid:
+
+```rust
+struct Foo {
+    a: &'static str,
+    b: bool = true,
+    c: i32,
+}
+```
+
+Invalid:
+
+```rust
+struct Foo {
+    a: &'static str,
+    b: Vec<bool> = Vec::new(),
+                   ^^^^^^^^^^
+                   // error: calls in field defaults are limited to struct and enum constructors
+    c: i32,
+}
+```
+
+Field defaults are sugar for the 'real' initialiser, where values for missing fields are added with the supplied default expression. 
+
+```rust
+let foo = Foo {
+    a: "Hello",
+    c: 42,
+};
+```
+
+Is equivalent to:
+
+```rust
+let foo = Foo {
+    a: "Hello",
+    b: true,
+    c: 42,
+};
+```
+
+When a field isn't supplied, and there is no default available, then the standard _missing field_ error applies.
+
+## Order of Precedence
+
+Supplied field values take precedence over field defaults:
+
+```rust
+// `b` is `false`, even though the field default is `true`
+let foo = Foo {
+    a: "Hello",
+    b: false,
+    c: 42,
+};
+```
+
+Supplied field values in functional updates take precedence over field defaults:
+
+```rust
+// `b` is `false`, even though the field default is `true`
+let foo = Foo {
+    a: "Hello",
+    c: 0,
+    ..Foo {
+        a: "Hello",
+        b: false,
+        c: 0
+    }
+};
+```
+
+The field default should only be used when the caller doesn't supply a value, to avoid unnecessarily assigning values.
+
+## Deriving Default
+
+When deriving `Default`, supplied field defaults are used instead of the type default. This is a feature of `#[derive(Default)]`.
+
+```rust
+#[derive(Default)]
+struct Foo {
+    a: &'static str,
+    b: bool = true,
+    c: i32,
+}
+
+// `b` is `true`, even though `bool::default()` is `false`
+let foo = Foo::default();
+```
```

Comment:
You should note that this means `derive(Default)` can be used more widely because only fields with default values need to implement `Default`